In [1]:
!pip install -e ..
import nextbike as nb
import os
from pathlib import Path
from sklearn.metrics import mean_squared_error, accuracy_score
import warnings
warnings.filterwarnings('ignore')

Obtaining file:///C:/Users/meikh/Dropbox/Programming%20Data%20Science/PDS2020_Herber
  Attempting uninstall: PDS2020-Herber
    Found existing installation: PDS2020-Herber 0.0.1
    Uninstalling PDS2020-Herber-0.0.1:
      Successfully uninstalled PDS2020-Herber-0.0.1
  Running setup.py develop for PDS2020-Herber


Using TensorFlow backend.


# Regression Task

In [2]:
# Declare name of the first csv-file
csv_name = "dortmund_test"
# Load dataset
path_a = Path(os.getcwd()).parent
path=os.path.join(path_a, f'data\input\{csv_name}_preprocessed_regression.csv')
df = nb.input_output.read_file(path)

In [3]:
# Adding missing locations
df['p_number_start7313']=0
df['p_number_start7332']=0
df['p_number_start7721']=0
df['p_number_start7722']=0
df['p_number_start7723']=0
df['p_number_start7724']=0
df['p_number_start7725']=0
df['p_number_start7726']=0
df['p_number_start7727']=0

Some locations from the train set are missing in the test set. As they are dummy coded, I can just add them with having the zero as value for all observations.

In [4]:
# Apply Random Forest Regression and calculate MSE
y_test = df['duration']
y_predict = nb.input_output.read_model('Random_Forest_Regression').predict(df)
print(mean_squared_error(y_test, y_predict))

7828.373970372764


While the mean squared error was about 5000 in the test set of the training data, it is here with 7828 by far worse for the Random Forest Regression.

In [5]:
# Calculate MSE for a Null Model
df['y_test_mean'] = y_test.mean()
print(mean_squared_error(y_test, df['y_test_mean']))

7018.824074794278


By just predicting trip duration with the mean of all durations also shows a higher MSE than in the training data, but is with a value of 7018 by far more useful than the Random Forest Regression. Conversely, this means that the Random Forest Regression is not only useless for this test data, but even misleading.

# Classification Task

In [6]:
# Load dataset
path_a = Path(os.getcwd()).parent
path=os.path.join(path_a, f'data\input\{csv_name}_preprocessed_classification.csv')
df = nb.input_output.read_file(path)

In [7]:
# Adding missing locations
df['p_number_start7313']=0
df['p_number_start7332']=0
df['p_number_start7721']=0
df['p_number_start7722']=0
df['p_number_start7723']=0
df['p_number_start7724']=0
df['p_number_start7725']=0
df['p_number_start7726']=0
df['p_number_start7727']=0

In [8]:
# Apply Random Forest Classification and calculate accuracy
y_test = df['towards_uni']
y_predict = nb.input_output.read_model('Random_Forest_Classification').predict(df)
print(accuracy_score(y_test, y_predict))

0.7280854403384857


Compared to the accuracy of the Random Forest Classification in the training data with 77.6%, the accuracy here is just 72.8%.

In [9]:
# Calculate accuracy for a Null Model
df['y_test_mean'] = round(y_test.mean())
print(accuracy_score(y_test, df['y_test_mean']))

0.7281195618794145


The accuracy of the null model applied to this test data is with 72.8% slightly higher than the one of the Random Forest Classifier.